### DISCUSSION TITLE
Using only LPS Nodes for Positioning of a Third-Party Device

### DISCUSSION THREAD

**Discussion by Kuipman** [bitcraze/discussions#1131] [Q&A]
Hello,

I'm a student researcher who has six Loco Positioning Nodes and wants to use these to build a positioning system based on TDoA3 for use with a third-party device such as a robot dog for thesis work. So far I've been able to load these up with firmware, set one node to a sniffer listening to three TDoA3 anchors, and use the python scripts in the lps-node-firmware repo to decode the TDoA3 messages into readable meter measurements between the anchors. Attached (hopefully the file works) is the output from running the raw binary from by the sniffer through the python scripts (numbers are in meters between anchors).
![Screenshot-63](https://github.com/bitcraze/discussions/assets/89929769/322f8c4f-ad9b-4638-a6ff-750b81ead22a)

I'm a bit stuck on where to go from here, I've been looking through the crazyflie repo and am having trouble finding code or a resource for using this data to calculate a 2D and eventually 3D position. Any pointers in the right direction would be much appreciated.

**Comment by knmcguire** [comment]
Hi! 

The most important files  for this are:
- [Kalman_core.c](https://github.com/bitcraze/crazyflie-firmware/blob/master/src/modules/src/kalman_core/kalman_core.c) (this is the code for the extended kalman filter
-  [mm_tdoa.c](https://github.com/bitcraze/crazyflie-firmware/blob/master/src/modules/src/kalman_core/mm_tdoa.c) (this is the measurement model code for the EKF for the [TDOA method](https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/functional-areas/loco-positioning-system/kalman_measurement_models/)).

Also we have [python bindings](https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/building-and-flashing/build/#build-python-bindings) of parts of the firmware, which includes the TDOA implementation of the loco positioning system. You can see a test for that right here that you can use as a reference: https://github.com/bitcraze/crazyflie-firmware/blob/master/test_python/test_kalman_core.py
